### DISCUSSION TITLE
storage.subscribe fires after atomWithStorage initializes against storage

### DISCUSSION THREAD

**Discussion by leoparis89** [pmndrs/jotai#1623] [Q&A]
## Summary

I'm using atomWithStorage like this
```javascript

import {atomWithStorage, createJSONStorage} from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';

const storage = createJSONStorage(() => AsyncStorage);

storage.subscribe = async (key, set) => {
       // This callback MUST be called before the app starts rendering, 
       // otherwise it will crash when running against oldShape

       if (key === "needsMigration") {
            const oldShape = await AsyncStorage.get("needsMigration")      
            set("needsMigration", update(oldShape))
       }
}

```

Is there a handle to do updates on the storage **before** my app renders  ?

Many thanks

**Comment by dai-shi** [comment]
I think it's tricky especially with AsyncStorage, because it's "async".

This is out of the scope of `atomWithStorage`. (In v2, we let the first render with the default value, and the update the value with storage value.)

You can create your own atom though.

```js
const getStorageValue = async () => ...
const setStorageValue = async (newValue) => ...

const myStorageAtom = atom(getStorageValue(), (get, set, newValue) => {
  set(myStorageAtom, newValue)
  setStorageValue(newValue)
})
```

It should work, but for TypeScript and best practice, we should probably create two atoms. Let me know if this approach works for you.

**Reply by EduardoLopes** [reply]
> In v2, we let the first render with the default value, and the update the value with storage value.

Why this decision? 

I just updated to v2 and my system checks the token (stored on local storage using `atomWithStorage`) in the first render to do actions and it's broken using v2

**Reply by dai-shi** [reply]
See: https://github.com/pmndrs/jotai/discussions/1737#discussioncomment-4857706

**Comment by leoparis89** [comment]
Thanks for your reply.
I would prefer avoid creating atoms just to get a handle on their data initialization.

I found the bellow solution to update the storage before the atoms are initialized.
Do you recommend this approach ?


```javascript

import {atomWithStorage, createJSONStorage} from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';

const storage = createJSONStorage(() => {
  return {
    ...AsyncStorage,
    getItem: async key => {
      // Update model in storage before atoms are initialized
      await doUpdatesInStorage(key);

      return AsyncStorage.getItem(key);
    },
  };
});

export const makeAtom = (key, content) =>
  atomWithStorage(key, content, storage);

```

**Reply by dai-shi** [reply]
Yes, it works. But, in v2, `delayInit: true` is default and not-configurable. Please add `delayInit: true` and see how it behaves.
